[Refactor] Use file factory to replace create file reader/writer

HappenLee opened the pull request:
1. Simplify code logic and improve abstraction
2. Fix the mem leak of raw pointer

# Proposed changes

Issue Number: close #xxx

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (No)
2. Has unit tests been added: (No Need)
3. Has document been added or modified: (No Need)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (Yes)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

yiguolei commented on be/src/exec/CMakeLists.txt:
```diff
@@ -92,25 +90,29 @@ set(EXEC_FILES
+    file/broker_writer.cpp
+    file/s3_reader.cpp
+    file/s3_writer.cpp
+    file/hdfs_reader_writer.cpp
```
Why not split to hdfs_reader.cpp and hdfs_writer.cpp as local, s3, broker ?

yiguolei commented on be/src/exec/broker_scanner.cpp:
```diff
@@ -158,53 +143,11 @@ Status BrokerScanner::open_file_reader() {
             _skip_lines = 2;
         }
     }
-    switch (range.file_type) {
```
I think it is better to use stream type not file type. 
FileType maybe parquet, orc, csv
StreamType could be S3Stream, HDFSStream, LocalStream

HappenLee replied:
Since the original code set the type in thrift, this name must be reserved for compatibility

yiguolei commented on be/src/exec/file/broker_reader.cpp:
```diff
@@ -15,7 +15,7 @@
 // specific language governing permissions and limitations
 // under the License.
 
-#include "exec/broker_reader.h"
```
create a be/src/io folder, and move all file reader and file writers to the IO folder.

HappenLee replied:
ok，I will do the work

HappenLee commented on be/src/exec/broker_scanner.cpp:
```diff
@@ -158,53 +143,11 @@ Status BrokerScanner::open_file_reader() {
+
+    RETURN_IF_ERROR(FileFactory::create_file_reader(range.file_type, _state->exec_env(),
+                                                    _broker_addresses, _params.properties, range,
+                                                    start_offset, _cur_file_reader));
```
origin code seems do not have the logic？

xy720 commented on be/src/exec/broker_scanner.cpp:
```diff
@@ -158,53 +143,11 @@ Status BrokerScanner::open_file_reader() {
-        FileReader* hdfs_file_reader;
-        RETURN_IF_ERROR(HdfsReaderWriter::create_reader(range.hdfs_params, range.path, start_offset,
-                                                        &hdfs_file_reader));
-        BufferedReader* file_reader = new BufferedReader(_profile, hdfs_file_reader);
```
Here, some file_reader is wrapped in a BufferedReader.

HappenLee replied:
yes，it should wrapped it. i will add the logic

HappenLee commented on be/src/exec/broker_scanner.h:
```diff
@@ -110,7 +110,7 @@ class BrokerScanner : public BaseScanner {
 
     // Reader
-    FileReader* _cur_file_reader;
+    std::shared_ptr<FileReader> _cur_file_reader;
```
not sharable

adonis0147 replied:
If the ownership of `_cur_file_reader` is unsharable, using `std::unique_ptr` is better accoding to [CppCoreGuidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#r21-prefer-unique_ptr-over-shared_ptr-unless-you-need-to-share-ownership). 
`std::shared_ptr` has some performance overhead. We should not mix the ownership and the access of the resource, otherwise, the `std::shared_ptr` will pollute lots of code.

HappenLee replied:
note the `std::shared_ptr<StreamLoadPipe> _stream_load_pipe;`, here use the shared_ptr, so now we can't not use other pointer to manage the lifetime of the `FileReader* _cur_file_reader;`. 
only the `shared_ptr`

HappenLee commented on be/src/exec/file/file_factory.h:
```diff
@@ -0,0 +1,74 @@
+    }
+
+private:
+    static Status _new_file_reader(TFileType::type type, ExecEnv* env,
+                                   const std::vector<TNetworkAddress>& broker_addresses,
+                                   const std::map<std::string, std::string>& properties,
+                                   const TBrokerRangeDesc& range, int64_t start_offset,
+                                   FileReader*& file_reader);
```
Yes, you are right. But please note that this is a `private` method, which itself should not directly called by users. There's no need to worry about what you said

adonis0147 replied:
I think there are two roles of users:
1. Ones who using the class.
2. Ones who develops the class.

It's ok for the users who using the class here due to the accessibility. However, for the ones who will modify the class, they should jump into the definition detail to figure the ownership out. That is inconvenient.

HappenLee replied:
I will add comment of this function to notice the develop of class to notice the lifetime of of `file_reader`

Commit:
[Refactor] Use file factory to replace create file reader/writer

1. Simplify code logic and improve abstraction
2. Fix the mem leak of raw pointer
